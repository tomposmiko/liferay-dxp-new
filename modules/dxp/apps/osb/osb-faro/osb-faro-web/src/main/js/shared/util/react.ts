import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
