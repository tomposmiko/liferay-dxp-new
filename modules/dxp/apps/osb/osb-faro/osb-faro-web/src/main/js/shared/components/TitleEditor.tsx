import autobind from 'autobind-decorator';
import ClayButton from '@clayui/button';
import ClayIcon from '@clayui/icon';
import getCN from 'classnames';
import React, {createRef} from 'react';
import {ENTER} from 'shared/util/key-constants';
import {withField} from 'shared/components/form';

interface ITitleEditorProps {
	editable?: boolean;
	inputName: string;
	onBlur?: (event) => void;
	onChange: (event) => void;
	placeholder?: string;
	value?: string;
}

export class TitleEditor extends React.Component<
	ITitleEditorProps,
	{editing: boolean}
> {
	static defaultProps = {
		editable: true
	};

	state = {
		editing: false
	};

	private _titleInput = createRef<HTMLInputElement>();

	@autobind
	handleBlur(event) {
		const {onBlur} = this.props;

		this.setState({editing: false});

		if (onBlur) {
			onBlur(event);
		}
	}

	@autobind
	handleEdit(event) {
		event.preventDefault();

		const {editable} = this.props;
		if (editable) {
			this.editing();
		}
	}

	@autobind
	handleKeyDown(event) {
		if (event.keyCode === ENTER) {
			this._titleInput.current.blur();
		}
	}

	@autobind
	handleKeyDownEdit(event) {
		event.preventDefault();

		if (event.keyCode === ENTER) {
			this.editing();
		}
	}

	@autobind
	editing() {
		this.setState(
			{
				editing: !this.state.editing
			},
			() => this._titleInput.current.select()
		);
	}

	render() {
		const {
			props: {editable, inputName, onChange, placeholder, value = ''},
			state: {editing}
		} = this;

		const rootClasses = getCN('title-editor-root', {editing});

		const inputClasses = getCN('title-input', {
			hide: !editing
		});

		const displayClasses = getCN('title-display', {
			hide: editing,
			'placeholder-display': !value
		});

		return (
			<div className={rootClasses}>
				<input
					className={inputClasses}
					name={inputName}
					onBlur={this.handleBlur}
					onChange={onChange}
					onKeyDown={this.handleKeyDown}
					placeholder={placeholder}
					ref={this._titleInput}
					required
					type='text'
					value={value}
				/>

				<div className={displayClasses}>
					<span
						className='title-value'
						onClick={this.handleEdit}
						onKeyDown={this.handleKeyDownEdit}
						role='button'
						tabIndex={0}
					>
						{value || placeholder}
					</span>

					{editable && (
						<ClayButton
							aria-label={Liferay.Language.get('edit')}
							borderless
							className='button-root'
							displayType='unstyled'
							onClick={this.handleEdit}
							outline
							size='sm'
						>
							<ClayIcon className='icon-root' symbol='pencil' />
						</ClayButton>
					)}
				</div>
			</div>
		);
	}
}

export default withField(({field: {name, ...otherFields}, ...otherProps}) => {
	const handleChange = event => {
		const {
			form: {setFieldValue}
		} = otherProps;

		setFieldValue(name, event.target.value);
	};

	return (
		<TitleEditor
			{...otherFields}
			{...otherProps}
			inputName={name}
			onChange={handleChange}
		/>
	);
});
