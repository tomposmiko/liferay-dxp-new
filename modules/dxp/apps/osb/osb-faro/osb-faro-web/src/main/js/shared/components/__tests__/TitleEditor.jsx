import React from 'react';
import {cleanup, fireEvent, render} from '@testing-library/react';
import {noop} from 'lodash';
import {TitleEditor} from '../TitleEditor';

jest.unmock('react-dom');

describe('TitleEditor', () => {
	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(<TitleEditor name='test' onChange={noop} />);
		expect(container).toMatchSnapshot();
	});

	it('should render with placeholder', () => {
		const placeholder = 'test placeholder';

		const {queryByPlaceholderText} = render(
			<TitleEditor
				name='test'
				onChange={noop}
				placeholder={placeholder}
			/>
		);

		expect(queryByPlaceholderText(placeholder)).toBeTruthy();
	});

	it('should render with value', () => {
		const value = 'foo';

		const {queryByText} = render(
			<TitleEditor name='test' onChange={noop} value={value} />
		);

		expect(queryByText(value)).toBeTruthy();
	});

	it('should render as editing', () => {
		const placeholder = 'test placeholder';

		const {container, getByPlaceholderText} = render(
			<TitleEditor
				name='test'
				onChange={noop}
				placeholder={placeholder}
			/>
		);

		const input = getByPlaceholderText(placeholder);

		expect(input).toHaveClass('hide');

		fireEvent.click(container.querySelector('.btn'));

		expect(input).not.toHaveClass('hide');
	});

	it('should not be editable if editable prop is false', () => {
		const {container} = render(
			<TitleEditor editable={false} name='test' onChange={noop} />
		);

		expect(container.querySelector('.btn')).toBeNull();
	});
});
