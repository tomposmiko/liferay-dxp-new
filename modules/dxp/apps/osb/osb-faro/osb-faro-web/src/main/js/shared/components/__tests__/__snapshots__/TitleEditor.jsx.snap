// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleEditor should render 1`] = `
<div>
  <div
    class="title-editor-root"
  >
    <input
      class="title-input hide"
      required=""
      type="text"
      value=""
    />
    <div
      class="title-display placeholder-display"
    >
      <span
        class="title-value"
        role="button"
        tabindex="0"
      />
      <button
        aria-label="Edit"
        class="button-root btn btn-outline-borderless btn-sm btn-outline-unstyled"
        type="button"
      >
        <svg
          class="lexicon-icon lexicon-icon-pencil icon-root"
          role="presentation"
        >
          <use
            xlink:href="#pencil"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
